rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,491 @@
+  | `id` | SDK identifier string as specified by [SDK-RFC #35](0035-rto.md#client--connection-ids) in section "Client / Connection IDs" |
+  | `node` | The hostname of the node as seen in configuration, and used by the SDK to establish connection. |
+  | `alt_node` | If alternative address is used, this field should be present |
+  | `bucket` | Name of the bucket (if present) |
+
+* `1` is a metric value. Each metric value increment represents a single
+  network operation (e.g. `GET_ALL_REPLICAS` should be counted as `1 +
+  number_of_replicas`).
+* `1695747260` is a metric timestamp, that represents time in milliseconds from
+   the Epoch (1970-01-01 00:00:00 UTC) to the moment of the generation this
+   response.
+
+#### Histograms
+
+    sdk_kv_retrieval_duration_seconds_bucket{le="0.001",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 24054
+    sdk_kv_retrieval_duration_seconds_bucket{le="0.010",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 33444
+    sdk_kv_retrieval_duration_seconds_bucket{le="0.100",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 100392
+    sdk_kv_retrieval_duration_seconds_bucket{le="0.5",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 129389
+    sdk_kv_retrieval_duration_seconds_bucket{le="1",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 133988
+    sdk_kv_retrieval_duration_seconds_bucket{le="2.5",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 139823
+    sdk_kv_retrieval_duration_seconds_bucket{le="+Inf",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 144320
+    sdk_kv_retrieval_duration_seconds_sum{agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 53423
+    sdk_kv_retrieval_duration_seconds_count{agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 144320
+
+Where
+
+* `24054` is a number of the points in the bucket
+* `sdk_kv_retrieval_duration_seconds_bucket` is a name of the histogram, where
+  `_bucket` is a special suffix, that signifies that the line should be
+  interpreted as a bucket of the histogram, and
+  `sdk_{service}_retrieval_duration_seconds` prefix describes particular metric
+  of the service. There are two special metrics associated with each histogram:
+  `_sum` and `_count` that represent the overall sum of all buckets, and total
+  number of the points in the histogram.
+* `le="0.001"` label is the upper bound of the bucket. All SDK must use hard
+  coded buckets to allow easy aggregation. The value would fall into all
+  buckets where upper bound is higher. The following buckets should be used
+  by the SDKs:
+
+<table>
+  <thead>
+    <tr>
+      <th>Name</th>
+      <th colspan="7">Buckets</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td>sdk_kv_retrieval</td>
```

Comment:
How do we represent transactions?

We decided elsewhere that we record the individual network ops rather than the overall compound op.  So I guess we don't have a top-level "sdk_transaction" op?

But do we have "sdk_kv_transaction_retrieval" ops?  I would think that, for simplicity, no?  We just use the existing "sdk_kv_retrieval" for each KV get or lookupIn done inside a transaction or transaction cleanup - including "odd balls" like ATR document fetches.  Similar for mutations (which will usually be "sdk_kv_mutation_durable").

@avsej do you agree?

Reply:
no top-level operations, and transactions will appear as a combinations of low-level operations.